File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static class Node {
        int x;
        int y;
        
        Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static int bfs(Node node, char[][] grid, boolean[][] visited) {
        visited[node.x][node.y] = true;
        Queue<Node> queue = new ArrayDeque<>();
        queue.add(node);

        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        int count = 0;

        while (!queue.isEmpty()) {
            Node current = queue.poll();

            for (int i=0; i<4; i++) {
                int nx = current.x + dx[i];
                int ny = current.y + dy[i];

                if (nx >= 0 && nx < grid.length && ny >= 0 && ny < grid[0].length) {
                    if (!visited[nx][ny] && grid[nx][ny] != 'X') {
                        visited[nx][ny] = true;
                        queue.add(new Node(nx, ny));
                        if (grid[nx][ny] == 'P') {
                            count++;
                        }
                    }
                }
            }
        }

        return count;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int result = 0;

        char[][] grid = new char[n][m];
        for (int i=0; i<n; i++) {
            String line = br.readLine();
            grid[i] = line.toCharArray();
        }

        boolean[][] visited = new boolean[n][m];

        outer:
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (grid[i][j] == 'I') {
                    result = bfs(new Node(i, j), grid, visited);
                    break outer;
                }
            }
        }

        System.out.println((result == 0) ? "TT" : result);
    }
}

File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.kt
/**
 * p21736 헌내기는 친구가 필요해
 */
fun bfs(x: Int, y: Int, graph: Array<CharArray>, visited: Array<BooleanArray>): Int {
    val queue = ArrayDeque<Pair<Int, Int>>()
    queue.addLast(x to y)
    visited[x][y] = true
    var count = 0

    val dirs = arrayOf(1 to 0, -1 to 0, 0 to 1, 0 to -1)
    while (queue.isNotEmpty()) {
        val current = queue.removeFirst()
        for (dir in dirs) {
            val nx = current.first + dir.first
            val ny = current.second + dir.second
            if (nx in graph.indices && ny in graph[0].indices && !visited[nx][ny] && graph[nx][ny] != 'X') {
                visited[nx][ny] = true
                queue.addLast(nx to ny)
                if (graph[nx][ny] == 'P') {
                    count++
                }
            }
        }
    }

    return count
}

fun main() {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val graph = Array(n) { readln().toCharArray() }
    val visited = Array(n) { BooleanArray(m) }
    var result = 0

    outer@ for (i in graph.indices) {
        for (j in graph[0].indices) {
            if (graph[i][j] == 'I') {
                result = bfs(i, j, graph, visited)
                break@outer
            }
        }
    }

    when (result) {
        0 -> println("TT")
        else -> println(result)
    }
}